[ScrollingTree mac] Fix scrolling when event region and scroll container are distinct

mdubet opened the pull request:
#### 2b7aa256562d1d2fa0bc22b53d9065a0d0556064
<pre>
[ScrollingTree mac] Fix scrolling when event region and scroll container are distinct
<a href="https://bugs.webkit.org/show_bug.cgi?id=258255">https://bugs.webkit.org/show_bug.cgi?id=258255</a>
<a href="https://rdar.apple.com/110954175">rdar://110954175</a>

Reviewed by Simon Fraser.

The current code (mac) assumes that the platform layer which reacts to
a scroll event is also the one containing the scroll container but it&apos;s not always the case
(like WPT example: when the scroll container has pointer-event:none but some child still reacts to event).

* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_hit_test_scroll_visible_descendant-expected.txt:
* Source/WebCore/page/scrolling/mac/ScrollingTreeMac.mm:
(ScrollingTreeMac::scrollingNodeForPoint):
* Source/WebCore/platform/graphics/cocoa/WebCoreCALayerExtras.h:
(WebCore::collectDescendantLayersAtPoint):
* Source/WebCore/platform/graphics/cocoa/WebCoreCALayerExtras.mm:
(WebCore::collectDescendantLayersAtPoint):
(WebCore::layersAtPointToCheckForScrolling):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
(WebKit::RemoteScrollingTreeMac::scrollingNodeForPoint):

Canonical link: <a href="https://commits.webkit.org/270094@main">https://commits.webkit.org/270094@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c45b02f874d968beec2235d62e04a7ad8775494b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/24537 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/3088 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/25805 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/26665 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/22556 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/24807 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/4735 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/523 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/22950 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/24782 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/2149 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21199 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/27252 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/1880 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/22125 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/28321 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/22376 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/22454 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26114 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/1808 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/138 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/3118 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5879 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/2266 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/2182 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/platform/graphics/cocoa/WebCoreCALayerExtras.h:
```diff
@@ -52,7 +52,7 @@
 
 using LayerAndPoint = std::pair<CALayer *, FloatPoint>;
-WEBCORE_EXPORT void collectDescendantLayersAtPoint(Vector<LayerAndPoint, 16>& layersAtPoint, CALayer *parent, CGPoint, const std::function<bool(CALayer *, CGPoint localPoint)>& pointInLayerFunction);
+WEBCORE_EXPORT void collectDescendantLayersAtPoint(Vector<LayerAndPoint, 16>& layersAtPoint, CALayer *parent, CGPoint, const std::optional<std::function<bool(CALayer *, CGPoint localPoint)>>& pointInLayerFunction = { });
```
`std::optional<std::function<>>` is rarely a good idea since std::function has a null state. You could just use `= nullptr` for the default value and then null-check the std::function before calling it.

mdubet replied:
TIL that `std::function` has a null state. 👍

cdumez commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
```diff
@@ -465,27 +465,56 @@ static bool layerEventRegionContainsPoint(CALayer *layer, CGPoint localPoint)
+            return rootScrollingNode;
+
+        for (auto [layer, point]: reversedLayersAtPointWithEventRegion) {
+            auto scrollingNodeForLayer = [&] (auto layer, auto point) -> ScrollingTreeNode* {
```
`RefPtr scrollingNodeForLayer`. Please avoid raw pointers in new code.

cdumez replied:
The lambda should likely return a `RefPtr< ScrollingTreeNode>` too since this is what it will have internally with the changes I suggest below.

cdumez commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
```diff
@@ -465,27 +465,56 @@ static bool layerEventRegionContainsPoint(CALayer *layer, CGPoint localPoint)
+            auto scrollingNodeForLayer = [&] (auto layer, auto point) -> ScrollingTreeNode* {
+                auto nodeID = scrollingNodeIDForLayer(layer);
+
+                auto* scrollingNode = nodeForID(nodeID);
```
RefPtr

cdumez commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
```diff
@@ -465,27 +465,56 @@ static bool layerEventRegionContainsPoint(CALayer *layer, CGPoint localPoint)
+                return nullptr;
+            };
+
+            if (auto scrollingNode = scrollingNodeForLayer(layer, point)) {
```
No curly brackets for one lines, also `RefPtr`

cdumez commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
```diff
@@ -465,27 +465,56 @@ static bool layerEventRegionContainsPoint(CALayer *layer, CGPoint localPoint)
+                startingLayer++;
+            auto parentLayer = startingLayer + 1;
+            while (parentLayer < reversedLayers.end()) {
+                if (auto scrollingNode = scrollingNodeForLayer(parentLayer->first, parentLayer->second))
```
RefPtr

nt1m commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
```diff
@@ -465,26 +465,54 @@ static bool layerEventRegionContainsPoint(CALayer *layer, CGPoint localPoint)
-            auto* scrollingNode = nodeForID(nodeID);
-            if (!is<ScrollingTreeScrollingNode>(scrollingNode))
-                continue;
+                RefPtr<ScrollingTreeNode> scrollingNode = nodeForID(nodeID);
```
```suggestion
                RefPtr scrollingNode = nodeForID(nodeID);
```

nt1m commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
```diff
@@ -465,26 +465,54 @@ static bool layerEventRegionContainsPoint(CALayer *layer, CGPoint localPoint)
+                return nullptr;
+            };
+
+            if (RefPtr<ScrollingTreeNode> scrollingNode = scrollingNodeForLayer(layer, point))
```
```suggestion
            if (RefPtr scrollingNode = scrollingNodeForLayer(layer, point))
```

smfr commented on LayoutTests/platform/mac-ventura-wk2/TestExpectations:
```diff
@@ -32,5 +32,8 @@ webgl/1.0.x/conformance/glsl/misc/shader-with-non-reserved-words-3-of-8.html [ P
 
 webkit.org/b/261444 [ Debug x86_64 ] http/tests/security/referrer-policy-header.html [ Pass Timeout ]
 
+# AsyncOverflowScrolling in enabled on Sonoma+
```
That comment isn't accurate. What you mean to say is "UI-side compositing is only enabled on Sonoma+".

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_hit_test_scroll_visible_descendant.html:
```diff
@@ -1,4 +1,4 @@
-<!doctype html>
+<!DOCTYPE html> <!-- webkit-test-runner [ AsyncOverflowScrollingEnabled=true ] -->
```
Please undo this change.

mdubet replied:
No, we need this to run EWS. It will be gone when we rebased after https://github.com/WebKit/WebKit/pull/19664 is merged

smfr commented on Source/WebCore/platform/graphics/cocoa/WebCoreCALayerExtras.h:
```diff
@@ -52,7 +52,24 @@
-WEBCORE_EXPORT void collectDescendantLayersAtPoint(Vector<LayerAndPoint, 16>& layersAtPoint, CALayer *parent, CGPoint, const std::function<bool(CALayer *, CGPoint localPoint)>& pointInLayerFunction);
+WEBCORE_EXPORT void collectDescendantLayersAtPoint(Vector<LayerAndPoint, 16>& layersAtPoint, CALayer *parent, CGPoint, const std::function<bool(CALayer *, CGPoint localPoint)>& pointInLayerFunction = { });
+
+auto layersAtPointToCheckForScrolling(auto layerEventRegionContainsPoint, auto scrollingNodeIDForLayer, const auto& layer, const auto& point, bool& hasAnyNonInteractiveScrollingLayers)
```
There's no reason to inline this function.

mdubet replied:
It was templated to avoid writing the long type of the function args...
It's implemented in the .cpp now

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
```diff
@@ -467,28 +467,47 @@ static bool layerEventRegionContainsPoint(CALayer *layer, CGPoint localPoint)
-            if (isScrolledBy(*this, nodeID, frontmostLayer)) {
-                LOG_WITH_STREAM(UIHitTesting, stream << "RemoteScrollingTreeMac " << this << " scrollingNodeForPoint " << point << " found scrolling node " << nodeID);
+            auto scrollingNodeForLayer = [&] (auto layer, auto point) -> RefPtr<ScrollingTreeNode> {
+                auto nodeID = scrollingNodeIDForLayer(layer);
```
Might as well throw in your `UNUSED_PARAM(point);` here too (since it's only used by logging).

nt1m commented on LayoutTests/TestExpectations:
```diff
@@ -6843,6 +6843,9 @@ imported/w3c/web-platform-tests/scroll-animations/scroll-timelines/two-animation
 imported/w3c/web-platform-tests/scroll-animations/view-timelines/range-boundary.html [ Skip ]
 
+# EWS can't run this test because AsyncOverflowScrolling is disabled
+imported/w3c/web-platform-tests/pointerevents/pointerevent_hit_test_scroll_visible_descendant.html [ Skip ]
```
That probably means no Interop points since wpt.fyi runs Ventura :(

mdubet replied:
AsyncOverflowScrolling is not Sonoma only (UI-side compositing is, but we do both code paths in this PR so we should be ok for Interop).

nt1m replied:
So why do we skip this test on Ventura? If this fails on Ventura, it'll fail on wpt.fyi

mdubet replied:
We skip everywhere here. (because for this to run on EWS, we first need to activate AsyncOverflow which will be done in another PR)

mdubet committed:
[ScrollingTree mac] Fix scrolling when event region and scroll container are distinct
https://bugs.webkit.org/show_bug.cgi?id=258255
rdar://110954175

Reviewed by Simon Fraser.

The current code (mac) assumes that the platform layer which reacts to
a scroll event is also the one containing the scroll container but it's not always the case
(like WPT example: when the scroll container has pointer-event:none but some child still reacts to event).

* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_hit_test_scroll_visible_descendant-expected.txt:
* Source/WebCore/page/scrolling/mac/ScrollingTreeMac.mm:
(ScrollingTreeMac::scrollingNodeForPoint):
* Source/WebCore/platform/graphics/cocoa/WebCoreCALayerExtras.h:
(WebCore::collectDescendantLayersAtPoint):
* Source/WebCore/platform/graphics/cocoa/WebCoreCALayerExtras.mm:
(WebCore::collectDescendantLayersAtPoint):
(WebCore::layersAtPointToCheckForScrolling):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
(WebKit::RemoteScrollingTreeMac::scrollingNodeForPoint):

Canonical link: https://commits.webkit.org/270094@main
